applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_cr_beam_element_3D2N.cpp

```diff
@@ -28,6 +29,76 @@ AdjointFiniteDifferenceCrBeamElement::~AdjointFiniteDifferenceCrBeamElement()
 {
 }
 
+void AdjointFiniteDifferenceCrBeamElement::Calculate(const Variable<Vector >& rVariable, Vector& rOutput, const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    // The particular soltuion of the influence function is rotated in global direction
+    if (rVariable == ADJOINT_PARTICULAR_DISPLACEMENT)
+    {
+        static constexpr int number_of_nodes = 2;
+        static constexpr int dimension = 3;
+        static constexpr unsigned int local_size = number_of_nodes * dimension;
+        static constexpr unsigned int element_size = local_size * 2;
+
+        KRATOS_ERROR_IF(rOutput.size() != element_size) << "Size of particular solution does not fit!" << std::endl;
+
+        CrBeamElement3D2N::Pointer p_primal_beam_element = dynamic_pointer_cast<CrBeamElement3D2N>(this->pGetPrimalElement());
+        BoundedMatrix<double, element_size, element_size> transformation_matrix = p_primal_beam_element->CalculateInitialLocalCS();
+        rOutput = prod(transformation_matrix, rOutput);
+    }
+
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferenceCrBeamElement::CalculateOnIntegrationPoints(const Variable<array_1d<double, 3 > >& rVariable,
+					      std::vector< array_1d<double, 3 > >& rOutput,
+					      const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    if (rVariable == ADJOINT_CURVATURE || rVariable == ADJOINT_STRAIN)
+    {
+        const double E = this->GetProperties()[YOUNG_MODULUS];
+        const double nu = this->GetProperties()[POISSON_RATIO];
+        const double G = E / (2.0 * (1.0 + nu));
+        const double A = this->GetProperties()[CROSS_AREA];
+        const double J = this->GetProperties()[TORSIONAL_INERTIA];
+        const double Iy = this->GetProperties()[I22];
+        const double Iz = this->GetProperties()[I33];
+
+        if (rVariable == ADJOINT_CURVATURE)
+        {
+            this->CalculateAdjointFieldOnIntegrationPoints(MOMENT, rOutput, rCurrentProcessInfo);
+
+            for (IndexType i = 0; i < rOutput.size(); ++i)
+            {
+                rOutput[i][0] *=  1.0 / (G * J);
+                rOutput[i][1] *= -1.0 / (E * Iy);
+                rOutput[i][2] *= -1.0 / (E * Iz);
+            }
+        }
+        else if (rVariable == ADJOINT_STRAIN)
+        {
+            this->CalculateAdjointFieldOnIntegrationPoints(FORCE, rOutput, rCurrentProcessInfo);
+
+            KRATOS_WARNING_IF("ADJOINT_STRAIN", (this->GetProperties().Has(AREA_EFFECTIVE_Y) || this->GetProperties().Has(AREA_EFFECTIVE_Z)))
+                        << "Not available for Timoschenko beam!" << std::endl;
+
+            for (IndexType i = 0; i < rOutput.size(); ++i)
+            {
+                rOutput[i][0] *= 1.0 / (E * A);
+                rOutput[i][1] *= 0.0;
+                rOutput[i][2] *= 0.0;
+            }
+        }
+    }
+    else
+        this->CalculateAdjointFieldOnIntegrationPoints(rVariable, rOutput, rCurrentProcessInfo);
```

Comment:
Is `CalculateAdjointFieldOnIntegrationPoints` a valid function for any variable except `ADJOINT_CURVATURE` and `ADJOINT_STRAIN`